source/base/parameter_acceptor.cc

```diff
@@ -0,0 +1,241 @@
+//-----------------------------------------------------------
+//
+//    Copyright (C) 2017 by the deal.II authors
+//
+//    This file is part of the deal.II library.
+//
+//    The deal.II library is free software; you can use it, redistribute
+//    it, and/or modify it under the terms of the GNU Lesser General
+//    Public License as published by the Free Software Foundation; either
+//    version 2.1 of the License, or (at your option) any later version.
+//    The full text of the license can be found in the file LICENSE at
+//    the top level of the deal.II distribution.
+//
+//-----------------------------------------------------------
+
+#include <deal.II/base/parameter_acceptor.h>
+#include <deal.II/base/utilities.h>
+#include <deal.II/base/revision.h>
+#include <deal.II/base/path_search.h>
+#include <boost/core/demangle.hpp>
+#include <fstream>
+
+DEAL_II_NAMESPACE_OPEN
+
+
+// Static empty class list
+std::vector<SmartPointer<ParameterAcceptor> > ParameterAcceptor::class_list;
+// Static parameter handler
+ParameterHandler ParameterAcceptor::prm;
+
+ParameterAcceptor::ParameterAcceptor(const std::string name) :
+  acceptor_id(class_list.size()),
+  section_name(name)
+{
+  SmartPointer<ParameterAcceptor> pt(this, boost::core::demangle(typeid(*this).name()).c_str());
+  class_list.push_back(pt);
+}
+
+
+ParameterAcceptor::~ParameterAcceptor()
+{
+  class_list[acceptor_id] = 0;
+}
+
+std::string ParameterAcceptor::get_section_name() const
+{
+  return (section_name != "" ? section_name : boost::core::demangle(typeid(*this).name()));
+}
+
+
+void
+ParameterAcceptor::initialize(const std::string filename,
+                              const std::string output_filename,
+                              const ParameterHandler::OutputStyle output_style_for_prm_format)
+{
+  declare_all_parameters();
+  if (filename != "")
+    {
+      // check the extension of input file
+      if (filename.substr(filename.find_last_of(".") + 1) == "prm")
```

Comment:
I think you should make sure that `.prm` is at the *end* of the filename

Reply:
that's exactly what this does. It finds the *last* ".", and takes whatever is left from the filename (the *last* extension).

Reply:
But it's also going to find `.prmabcdef`.

Reply:
in which case it will correctly fail, because we don't support file types `.prmabcdef`. no?

Reply:
Well, it won't fail -- if I give you a filename `file.prmabc`, then you will end up in this `if` branch here even though it should be rejected as an unrecognized extension.

Reply:
Well... I think you are confused by `substr`. The check I'd be doing is the following:
~~~
std::string(".prmabc") == ".prm"
~~~
which is (correctly so) false.

Reply:
Oh -- you are saying that `string::substring(pos)` just takes everything starting at `pos`? I guess that makes sense now that you say it. Sorry for leading you around in a circle :-(

Reply:
~~~
basic_string substr( size_type pos = 0,  size_type count = npos ) const;
~~~
> Returns a substring [pos, pos+count). If the requested substring extends past the end of the string, or if count == npos, the returned substring is [pos, size()).

:)